source/extensions/common/dynamic_forward_proxy/cluster_store.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include "envoy/upstream/upstream.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace DynamicForwardProxy {
+
+class DfpCluster {
+public:
+  virtual ~DfpCluster() = default;
+
+  /**
+   * @return the cluster enabled subCluster configuration or not.
+   */
+  virtual bool enableSubCluster() const PURE;
+
+  /**
+   * Create a full cluster config for the cluster_name, with the specified host and port.
+   * @return true and nullptr when the subCluster with the specified cluster_name already
+   *         created, or true and the created cluster config when it not exists and not
+   *         reach the limitation of max_sub_clusters, otherwise, return false and nullptr.
+   */
+  virtual std::pair<bool, absl::optional<envoy::config::cluster::v3::Cluster>>
+  createSubClusterConfig(const std::string& cluster_name, const std::string& host,
+                         const int port) PURE;
+
+  /**
+   * Update the last used time of the subCluster with the specified cluster_name.
+   * @return true if the subCluster is existing.
+   */
+  virtual bool touch(const std::string& cluster_name) PURE;
+};
+
+using DfpClusterSharedPtr = std::shared_ptr<DfpCluster>;
+using DfpClusterWeakPtr = std::weak_ptr<DfpCluster>;
+
+class DFPClusterStore {
+public:
+  // Load the dynamic forward proxy cluster from this store.
+  static DfpClusterSharedPtr load(std::string cluster_name);
+
+  // Save the dynamic forward proxy cluster into this store.
+  static void save(const std::string cluster_name, DfpClusterSharedPtr cluster);
+
+  // Remove the dynamic forward proxy cluster from this store.
+  static void remove(std::string cluster_name);
+
+private:
+  using ClusterMapType = absl::flat_hash_map<std::string, DfpClusterWeakPtr>;
+  struct ClusterStoreType {
+    ClusterMapType map_ ABSL_GUARDED_BY(mutex_);
+    absl::Mutex mutex_;
+  };
+
+  static ClusterStoreType& getClusterStore() { MUTABLE_CONSTRUCT_ON_FIRST_USE(ClusterStoreType); }
+};
```

Comment:
We shouldn't be using statics for this. Use SingletonManager, which I think may solve your shutdown problem.

Reply:
changed to use `SingletonManager`, but still need `tls_.isShutdown` in `Cluster::~Cluster`.

without the `tls_.isShutdown()`, test will hit `assert failure: !shutdown_` while removing cluster.
And here is the full log:
```
[2023-05-08 07:01:15.527][12][info][testing] [test/integration/server.cc:255] stopping integration test server
[2023-05-08 07:01:15.527][233][info][main] [source/server/server.cc:1005] shutting down server instance

... more logs ...

[2023-05-08 07:01:15.529][233][debug][upstream] [source/extensions/clusters/dynamic_forward_proxy/cluster.cc:61] cluster='DFPCluster:sni.lyft.com:443' removing from cluster_map & cluster manager
[2023-05-08 07:01:15.530][233][debug][init] [source/common/init/watcher_impl.cc:31] ClusterImplBase destroyed
[2023-05-08 07:01:15.530][233][debug][init] [source/common/init/watcher_impl.cc:31] init manager Cluster DFPCluster:sni.lyft.com:443 destroyed
[2023-05-08 07:01:15.530][233][debug][upstream] [source/common/upstream/cluster_manager_impl.cc:808] removing cluster DFPCluster:sni.lyft.com:443
[2023-05-08 07:01:15.530][233][critical][assert] [source/common/thread_local/thread_local_impl.cc:152] assert failure: !shutdown_.
```

maybe using `tls_.isShutdown` is acceptable here?